File: erlproc.erl
%% @author J David Eisenberg <jdavid.eisenberg@gmail.com>
%% @doc Create graphics from Erlang using Processing's core functions.
%% @copyright 2013 by J David Eisenberg under Erlang Public License (same as Erlang)
%% @version 0.1

%% The system works on the basis of three processes:
%% 1) The hub process (this module), which communicates with the other
%%    two processes.
%% 2) The Java process. It gets a list of things to draw from the hub,
%%    and sends the drawing environment to the hub.
%% 3) The sketch process. It sends the hub things that it wants to draw,
%%    and awaits the hub's response. This is necessary to make sure
%%    that calls to the hub stay in order.

-module(erlproc).
-export([
  sketch/2,
  hub/2, hub_loop/5,
  redraw/0,
  no_loop/0,
  point/1,
  line/1,
  rect/1,
  ellipse/1,
  triangle/1,
  quad/1,
  arc/1,
  fill/1, no_fill/0,
  stroke/1, no_stroke/0,
  stroke_weight/1,
  background/1,
  smooth/0,
  mouse/0,
  pmouse/0,
  mouse_pressed/0,
  mouse_button/0,
  rect_mode/1,
  ellipse_mode/1
]).


%% @doc Start a sketch with the given dimensions and name of a module
%% that contains a setup() and draw() function.

-spec(sketch(atom(), [float()]) -> atom()).

sketch(ModuleName, Dimensions) ->
  %% Set up environment.
  
  %G = [{mouseX, 0}, {mouseY, 0}, {pmouseX, 0}, {pmouseY, 0},
  %  {mousePressed, false}, {mouseButton, none}, {width, 0},
  %  {height, 0}],

  %%If already running the hub process, do nothing, otherwise spawn it.
  case lists:member(hub, registered()) of
    true -> ok;
    false ->
      Pid = spawn_link(?MODULE, hub, [ModuleName, Dimensions]),
      register(hub, Pid)
      % io:format(standard_error, "erlproc registered as ~p~n", [Pid])
  end.

hub(Module, Dimensions) ->
  process_flag(trap_exit, true),
  %% Kill the Java process if it's running.
  case lists:member(java_process, registered()) of
    true ->
      % exit(whereis(erlproc_java), kill),
      port_close(java_process);
    false -> ok
  end,

  Port = open_port({spawn, "java -cp .:/usr/local/lib/erlang/lib/jinterface-1.5.8/priv/OtpErlang.jar:core.jar ErlProc"},
    [{line, 256}]),
  link(Port), % link to it
  register(java_process, Port), % and register it
	
  %% Although I could communicate between Erlang and Java via
  %% process name / node tuples, I felt it would be better to use
  %% Pids. Thus, the following handshaking kabuki to exchange
  %% Pids.

  %% First, I send the name of this node to Java via the port
  port_command(Port, list_to_binary(atom_to_list(node()) ++ "\n" )),
  
  %% Java will send me back its Pid and node name
  receive 
    {kabuki, JavaPid, _JavaNodeName} ->
      % io:format(standard_error, "Erlang gets Java's pid: ~p and name ~p~n",
      %  [JavaPid, _JavaNodeName]),
      % io:format(standard_error, "Sending message to Java~n", []),
      
      %% and send Java a message which has my pid, and the sketch dimensions
      JavaPid ! {kabuki, self(), Dimensions},  
      
      %% and wait for Java to respond to me
      receive
        kabuki_complete ->
          % io:format(standard_error, "Erlang reports kabuki complete~n", []),
     
          %% Start the hub's main loop
          hub_loop(Module, undefined, JavaPid, [], []) 
      end;
      
    _X ->
      io:format(standard_error,
        "Erlang received unexpected handshake message ~p~n", [_X])
  end.

hub_loop(SketchModule, SketchPid, JavaPid, Environment, DrawList) ->
  receive
    %% Java sends me its environment; I acknowledge.
    {environment, NewEnvironment} ->
      % io:format(standard_error, "Received environment ~p~n", [NewEnvironment]),
      JavaPid ! ok,
      hub_loop(SketchModule, SketchPid, JavaPid, NewEnvironment, DrawList);
    
    setup ->
      % io:format(standard_error, "From java: setup/0 on ~p~n", [SketchModule]),
      SetupFcnPid = spawn_link(SketchModule, setup, []),
      % io:format(standard_error, "Spawned setup at PID ~p~n", [SetupFcnPid]),
      hub_loop(SketchModule, SetupFcnPid, JavaPid, Environment, DrawList);
    
    draw ->
      % io:format(standard_error, "From java: draw/0~n", []),
      DrawFcnPid = spawn_link(SketchModule, draw, []),
      % io:format(standard_error, "Spawned draw at pid ~p~n", [DrawFcnPid]),
      hub_loop(SketchModule, DrawFcnPid, JavaPid, Environment, DrawList);
    
    {draw_cmd, Command}  when SketchPid /= undefined ->
      % io:format(standard_error, "Add ~p to draw list~n", [Command]),
      NewDrawList = [{Command, []} | DrawList],
      SketchPid ! ok,
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, NewDrawList);

    {draw_cmd, Command, Data} when SketchPid /= undefined ->
      % io:format(standard_error, "Add ~p ~p to draw list~n", [Command, Data]),
      NewDrawList = [{Command, Data} | DrawList],
      SketchPid ! ok,
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, NewDrawList);
    
    {draw_request, mouse} ->
      {_Mx, X} = lists:keyfind(mouseX, 1, Environment),
      {_My, Y} = lists:keyfind(mouseY, 1, Environment),
      SketchPid ! [X, Y],
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, DrawList);

    {draw_request, pmouse} ->
      {_Px, X} = lists:keyfind(pmouseX, 1, Environment),
      {_Py, Y} = lists:keyfind(pmouseY, 1, Environment),
      SketchPid ! [X, Y],
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, DrawList);

    {draw_request, mouse_pressed} ->
      {_K, Status} = lists:keyfind(mousePressed, 1, Environment),
      SketchPid ! Status,
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, DrawList);

    {draw_request, mouse_button} ->
      {_K, Button} = lists:keyfind(mouseButton, 1, Environment),
      SketchPid ! Button,
      hub_loop(SketchModule, SketchPid, JavaPid, Environment, DrawList);

    {'EXIT', SketchPid, normal} ->
      % io:format(standard_error, "Sketch function ~p exits normally.~n",
      % [SketchPid]),
      JavaPid ! {execute_commands, DrawList},
      hub_loop(SketchModule, undefined, JavaPid, Environment, []);
    
    {'EXIT', JavaPid, _} ->
      io:format(standard_error, "Java program exits. Ending erlang loop~n", []);

    X ->
      io:format(standard_error, "Hub loop gets unknown message ~p~n", [X]),
      exit(JavaPid, kill)
  end.

await() ->
  receive
    _X -> % io:format(standard_error, "Acknowledgment: ~p~n", [_X]),
      _X
  end.
  
%% These are drawing functions that will be called from the sketch.
%% Each one tells the hub to add a command to the draw list,
%% and then waits for a reply from the hub before continuing.

redraw() ->
  hub ! {draw_cmd, redraw},
  await().
  
no_loop() ->
  hub ! {draw_cmd, noLoop},
  await().

point(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, point, Coords},
  await().

line(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, line, Coords},
  await().

rect(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, rect, Coords},
  await().

rect_mode(Mode) when is_atom(Mode) ->
  hub ! {draw_cmd, rect_mode, Mode},
  await().

triangle(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, triangle, Coords},
  await().

quad(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, quad, Coords},
  await().

arc(Coords) when is_list(Coords) ->
  hub ! {draw_cmd, arc, Coords},
  await().

ellipse(Coords) when is_list(Coords) -> 
  hub ! {draw_cmd, ellipse, Coords},
  await().
  
ellipse_mode(Mode) when is_atom(Mode) ->
  hub ! {draw_cmd, ellipse_mode, Mode},
  await().

background(ColorList) when is_list(ColorList) ->
  hub ! {draw_cmd, background, ColorList},
  await().

smooth() ->
  hub ! {draw_cmd, smooth},
  await().

fill(ColorList) when is_list(ColorList) ->
  hub ! {draw_cmd, fill, ColorList},
  await().

no_fill() ->
  hub ! {draw_cmd, noFill},
  await().

stroke(ColorList) when is_list(ColorList) ->
  hub ! {draw_cmd, stroke, ColorList},
  await().

no_stroke() ->
  hub ! {draw_cmd, noStroke},
  await().

stroke_weight(W) ->
  hub ! {draw_cmd, strokeWeight, [W]},
  await().

mouse() ->
  hub ! {draw_request, mouse},
  receive
    [X, Y] -> [X, Y];
    _Other -> [0.0, 0.0]
  end.

pmouse() ->
  hub ! {draw_request, pmouse},
  receive
    [X, Y] -> [X, Y];
    _Other -> [0.0, 0.0]
  end.

mouse_pressed() ->
  hub ! {draw_request, mouse_pressed},
  receive
    Status -> Status
  end.

mouse_button() ->
  hub ! {draw_request, mouse_button},
  receive
    Button -> Button
  end.

File: ErlProc.java
/*
  Interface Processing's core functions to Erlang
  Copyright 2013 by J. David Eisenberg
  Licensed under LGPL version 2.1 (same as Processing's core)
*/
import com.ericsson.otp.erlang.*;
import processing.core.*;
import java.awt.BorderLayout;
import java.io.BufferedInputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.swing.JFrame;

public class ErlProc
{
  OtpSelf self;
  OtpConnection connection;

  OtpErlangPid erlPid;
  String erlNodeName;
  
  String erlFunctionName;

  ProcessingFrame frame;
  Embedded embed;

  public ErlProc()
  {
  }
  
  /**
   * Read a line from standard input. Output from Erlang may not be
   * buffered, so I do not want to make a BufferedReader based on System.in.
   */
  private String getLine()
  {
    byte[] buf = new byte[256];
    int ch = 0;
    int i = 0;
    try {
      ch = System.in.read();
      while (ch > 0 && ch != '\n' && i < 256)
      {
        if (ch >= 32) // skip control characters
        {
          buf[i++] = (byte) ch;
        }
        ch = System.in.read();
      }
    }
    catch (Exception e)
    {
      i = 0; // bail; return null string
    }
    return new String(buf, 0, i);
  }

  /**
   * Establish a connection with an Erlang node. Start by receiving
   * the Erlang node's name via the port.
   */
  public void go()
  {
    String str;
    OtpPeer erlangPeer;
    OtpMsg message;
    OtpErlangTuple tuple;
    OtpErlangAtom command;
    OtpErlangList dimensionList;
    float [] dimensions;
    
    try
    {
      OtpErlangObject [] items = new OtpErlangObject[3];
      OtpErlangObject obj;
      
      /*
        The Erlang node name comes via the port
      */
      erlNodeName = getLine();
      
      /* Create a connection to Erlang */
      self = new OtpSelf("java_process");
      erlangPeer = new OtpPeer(erlNodeName);
      connection = self.connect(erlangPeer);
      // System.err.println("Connected to " + erlNodeName 
      //  + " " + connection + "\r");
      
      /*
        Send a message to the Erlang process by name;
        it will have my Pid and node name in it.
      */
      // System.err.println("Java's pid is " + connection.self().pid() + "\r");
      items[0] = new OtpErlangAtom("kabuki");
      items[1] = connection.self().pid();
      items[2] = new OtpErlangString("java_process");
      connection.send("hub", new OtpErlangTuple(items));
      
      /*
        Erlang sends me a message from which I can get its Pid
        and the sketch dimensions
      */
      message = connection.receiveMsg();
      obj = message.getMsg();
      // System.err.println("Java gets message " + obj + "\r");
      tuple = (OtpErlangTuple) obj;
      command = (OtpErlangAtom) tuple.elementAt(0);
 
      /* And I send back a message to let Erlang know handshaking is finished */
      if (command.atomValue().equals("kabuki"))
      {
        int width;
        int height;
        
        erlPid = (OtpErlangPid) tuple.elementAt(1);
        dimensions = getFloatList(tuple.elementAt(2), 2);
        width = (int) dimensions[0];
        height = (int) dimensions[1];
        
        // System.err.println("Java gets Erlang's Pid: " + erlPid +
        //  " Dimensions: " + dimensionList  + "\r");
        connection.send(erlPid, new OtpErlangAtom("kabuki_complete"));
        
        // open the frame, and start the embedded PApplet
        frame = new ProcessingFrame(connection, width, height);
        embed = frame.embed;
        frame.setResizable(true);
        frame.setSize(width + 20, height + 20);
        frame.setVisible(true);
        embed.init();
      }
      else
      {
        // System.err.println("Java gets unknown command " + command.atomValue()
        //  + "\r");
      }
    }
    catch (Exception e)
    {
      // System.err.println("ERR" + e);
      e.printStackTrace();
    }
  }
  
  /**
   * Retrieve a list of floating point numbers from an Erlang list.
   */
  private float[] getFloatList(OtpErlangObject obj, int maxItems)
  {
    float[] result = null;
    
    /*
      Erlang represents a list of integers < 255 as a string.
      If I get a string, I have to convert each of those integers
      to a float.
    */
    if (OtpErlangString.class.isInstance(obj))
    {
      String str = ((OtpErlangString) obj).stringValue();
      int[] codePoints = OtpErlangString.stringToCodePoints(str);
      result = new float[str.length()];
      for (int i = 0; i < Math.min(maxItems, str.length()); i++)
      {
        result[i] = codePoints[i];
      }
    }
    else if (OtpErlangList.class.isInstance(obj)) // normal Erlang list
    {
      OtpErlangList list = (OtpErlangList) obj;
      int nItems = Math.min(maxItems, list.arity());
      result = new float[nItems];
      /*
        Coerce doubles and integers into floats. An OtpErlangFloat is
        a subclass of OtpErlangDouble, and an OtpErlangInt is a subclass of
        OtpErlangLong, so that's all I have to check for.
      */
      for (int i = 0; i < nItems; i++)
      {
        try
        {
          if (OtpErlangDouble.class.isInstance(list.elementAt(i)))
          {
            result[i] = ((OtpErlangDouble) list.elementAt(i)).floatValue();
          }
          else if (OtpErlangLong.class.isInstance(list.elementAt(i)))
          {
            result[i] = ((OtpErlangLong) list.elementAt(i)).intValue();
          }
          else
          {
            result[i] = 0.0F; // if I can't figure out what it is, bail out.
          }
        }
        catch (Exception e)
        {
          e.printStackTrace(System.err);
          result[i] = 0.0F; // likewise if coercion fails
        }
      }
    }
    return result;
  }
  
  /**
   * Accept messages from Erlang to draw to the sketch. This is where
   * all the work really happens.
   */
  private void executeDrawingCommands(OtpErlangList drawingList)
  {
    boolean exit = false;
    OtpErlangTuple tuple;
    
    String command;
    float[] coords;
    int nItems = drawingList.arity();
    
    for (int i = nItems -1 ; i >= 0; i--)
    {
      tuple = (OtpErlangTuple) drawingList.elementAt(i);
      command = ((OtpErlangAtom) tuple.elementAt(0)).atomValue();
        
      // System.err.println("Executing: " + command + "\r");
      if (command.equals("redraw"))
      {
        embed.redraw();
        embed.noLoop();
      }
      else if (command.equals("noLoop"))
      {
        embed.noLoop();
      }
      else if (command.equals("fill"))
      {
        doFill(tuple.elementAt(1));
      }
      else if (command.equals("noFill"))
      {
        embed.noFill();
      }
      else if (command.equals("stroke"))
      {
        doStroke(tuple.elementAt(1));
      }
      else if (command.equals("nostroke"))
      {
        embed.noStroke();
      }
      else if (command.equals("background"))
      {
        doBackground(tuple.elementAt(1));
      }
      else if (command.equals("line"))
      {
        coords = getFloatList(tuple.elementAt(1), 4);
        if (coords.length == 4)
        {
          embed.line(coords[0], coords[1], coords[2], coords[3]);
        }
      }
      else if (command.equals("rect"))
      {
        coords = getFloatList(tuple.elementAt(1), 4);
        if (coords.length == 4)
        {
          embed.rect(coords[0], coords[1], coords[2], coords[3]);
        }
      }
      else if (command.equals("ellipse"))
      {
        coords = getFloatList(tuple.elementAt(1), 4);
        if (coords.length == 4)
        {
          embed.ellipse(coords[0], coords[1], coords[2], coords[3]);
        }
      }
      else if (command.equals("triangle"))
      {
        coords = getFloatList(tuple.elementAt(1), 6);
        if (coords.length == 6)
        {
          embed.triangle(coords[0], coords[1], coords[2], coords[3],
            coords[4], coords[5]);
        }
      }
      else if (command.equals("quad"))
      {
        coords = getFloatList(tuple.elementAt(1), 8);
        if (coords.length == 8)
        {
          embed.quad(coords[0], coords[1], coords[2], coords[3],
            coords[4], coords[5], coords[6], coords[7]);
        }
      }
    }
  }

  private void doFill(OtpErlangObject params)
  {
    float[] values = getFloatList(params, 4);
    switch (values.length)
    {
      case 1: embed.fill(values[0]); break;
      case 2: embed.fill(values[0], values[1]); break;
      case 3: embed.fill(values[0], values[1], values[2]); break;
      case 4: embed.fill(values[0], values[1], values[2], values[3]); break;
      default: break;
    }
  }

  private void doBackground(OtpErlangObject params)
  {
    float[] values = getFloatList(params, 4);
    switch (values.length)
    {
      case 1: embed.background(values[0]); break;
      case 2: embed.background(values[0], values[1]); break;
      case 3: embed.background(values[0], values[1], values[2]); break;
      case 4: embed.background(values[0], values[1], values[2], values[3]); break;
      default: break;
    }
  }

  private void doStroke(OtpErlangObject params)
  {
    float[] values = getFloatList(params, 4);
    switch (values.length)
    {
      case 1: embed.stroke(values[0]); break;
      case 2: embed.stroke(values[0], values[1]); break;
      case 3: embed.stroke(values[0], values[1], values[2]); break;
      case 4: embed.stroke(values[0], values[1], values[2], values[3]); break;
      default: break;
    }
  }

  class Embedded extends PApplet {
    OtpErlangObject obj;
    OtpErlangList drawingList;
    OtpConnection connection;
    int width;
    int height;
    
    public Embedded(OtpConnection connection, int width, int height)
    {
      super();
      this.connection = connection;
      this.width = width;
      this.height = height;
    }
    
    private OtpErlangTuple createTuple(String atom, int value)
    {
      OtpErlangObject[] elements = new OtpErlangObject[2];
      elements[0] = new OtpErlangAtom(atom);
      elements[1] = new OtpErlangFloat(value);
      return new OtpErlangTuple(elements);
    }

    private OtpErlangTuple getEnvironment()
    {
      OtpErlangTuple [] items = new OtpErlangTuple[8];
      OtpErlangObject [] elements = new OtpErlangObject[2];
      OtpErlangTuple result;
      
      items[0] = createTuple("mouseX", this.mouseX);
      items[1] = createTuple("mouseY", this.mouseY);
      items[2] = createTuple("pmouseX", this.pmouseX);
      items[3] = createTuple("pmouseY", this.pmouseY);
      
      elements[0] = new OtpErlangAtom("mousePressed");
      elements[1] = new OtpErlangBoolean(this.mousePressed);
      items[4] = new OtpErlangTuple(elements);
      
      elements[0] = new OtpErlangAtom("mouseButton");
      if (this.mouseButton == PApplet.LEFT)
      {
        elements[1] = new OtpErlangAtom("left");
      }
      else if (this.mouseButton == PApplet.RIGHT)
      {
        elements[1] = new OtpErlangAtom("right");
      }
      else if (this.mouseButton == PApplet.CENTER)
      {
        elements[1] = new OtpErlangAtom("center");
      }
      else
      {
        elements[1] = new OtpErlangAtom("none");
      }
      items[5] = new OtpErlangTuple(elements);

      items[6] = createTuple("width", this.width);
      items[7] = createTuple("height", this.height);
      
      elements[0] = new OtpErlangAtom("environment");
      elements[1] = new OtpErlangList(items);
      result = new OtpErlangTuple(elements);
      return(result);
    }

    public void setup()
    {
      OtpErlangAtom atom;
      OtpErlangTuple tuple;
      
      size(width, height);
      try
      {
        // System.err.println("Java is sending environment\r");
        connection.send(erlPid, getEnvironment());
        atom = (OtpErlangAtom) connection.receiveMsg().getMsg();
        if (atom.atomValue().equals("ok"))
        {
          connection.send(erlPid, new OtpErlangAtom("setup"));
          tuple  = (OtpErlangTuple) connection.receiveMsg().getMsg();
          // System.err.println("Java told: " +
          //  ((OtpErlangAtom) tuple.elementAt(0)).atomValue() + "\r");
          executeDrawingCommands((OtpErlangList) tuple.elementAt(1));
        }
      }
      catch (Exception e)
      {
        // System.err.println("Cannot setup.\r");
        e.printStackTrace(System.err);
        System.exit(1);
      }
    }
  
    public void draw()
    {
      OtpErlangAtom atom;
      OtpErlangTuple tuple;
      
      try
      {
        connection.send(erlPid, getEnvironment());
        atom = (OtpErlangAtom) connection.receiveMsg().getMsg();
        if (atom.atomValue().equals("ok"))
        {
          connection.send(erlPid, new OtpErlangAtom("draw"));
          tuple  = (OtpErlangTuple) connection.receiveMsg().getMsg();
          // System.err.println("Java told: " +
          //  ((OtpErlangAtom) tuple.elementAt(0)).atomValue() + "\r");
          executeDrawingCommands((OtpErlangList) tuple.elementAt(1));
        }
      }
      catch (Exception e)
      {
        // System.err.println("Cannot draw.\r");
        e.printStackTrace(System.err);
        System.exit(1);
      }
    }
  }

  class ProcessingFrame extends JFrame {
  
    public Embedded embed;
      
    public ProcessingFrame(OtpConnection connection, int width, int height) {
      super("Processing from Erlang");
      this.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
      setLayout(new BorderLayout());
           
      Embedded embed = new Embedded(connection, width, height);
      add(embed, BorderLayout.CENTER);
      this.embed = embed;
    }
  }

  public static void main(String [] args)
  {
    ErlProc app = new ErlProc();
    app.go();    
  }

}
